### PR TITLE
ST-964 Contact Preference Type Check

### PR DESCRIPTION
**Pull request by ElifDenizEsea**
**Before creating a pull request make sure that:**

- [x] commit messages are meaningful and follow good commit message guidelines
- [x] README and other documentation has been updated / added (if needed)
- [x] tests have been updated / new tests has been added (if needed)

### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ST-964

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/sptribs/ciccase/model/ContactPreferenceType.java [hunk 1]
```diff
@@ -10,15 +10,18 @@
...
     POST("Post", "Post");
     private String type;
     private final String label;
+
     public boolean isEmail() {
         return EMAIL.name().equalsIgnoreCase(this.name());
     }
 
+    public boolean isPost() {
+        return POST.name().equalsIgnoreCase(this.name());
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/sptribs/ciccase/model/ContactPreferenceType.java:25]: Enums are singleton pattern - wouldn't `return POST == this;` work just as well? == is superior when checking enum equality, as it caters for nulls.

**Review by osheaf01 — CHANGES_REQUESTED** [review]
LGTM, just one query about the equality check?

File: src/main/java/uk/gov/hmcts/sptribs/common/event/page/ApplicantDetails.java [hunk 2]
```diff
@@ -41,7 +42,9 @@ public AboutToStartOrSubmitResponse<CaseData, State> midEvent(
         final CaseData data = details.getData();
         final List<String> errors = new ArrayList<>();
 
-        if (null != data.getCicCase() && null != data.getCicCase().getApplicantAddress()) {
+        if (null != data.getCicCase()
+            && data.getCicCase().getApplicantContactDetailsPreference().equals(ContactPreferenceType.POST)
+            && null != data.getCicCase().getApplicantAddress()) {
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/sptribs/common/event/page/ApplicantDetails.java:47]: .equals defers to ==, I believe, but == is null-safe and it also checks type compatibility at compile time.
I'd also put the entity you know exists on the LHS, so

`ContactPreferenceType.POST == data.getCicCase().getApplicantContatDetailsPreference()` ?

File: src/main/java/uk/gov/hmcts/sptribs/common/event/page/RepresentativeDetails.java [hunk 3]
```diff
@@ -42,7 +43,9 @@ public AboutToStartOrSubmitResponse<CaseData, State> midEvent(
         final CaseData data = details.getData();
         final List<String> errors = new ArrayList<>();
 
-        if (null != data.getCicCase() && null != data.getCicCase().getRepresentativeAddress()) {
+        if (null != data.getCicCase()
+            && data.getCicCase().getRepresentativeContactDetailsPreference().equals(ContactPreferenceType.POST)
+            && null != data.getCicCase().getRepresentativeAddress()) {
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/sptribs/common/event/page/RepresentativeDetails.java:48]: As for comment above.

File: src/main/java/uk/gov/hmcts/sptribs/common/event/page/SubjectDetails.java [hunk 4]
```diff
@@ -35,7 +36,9 @@ public AboutToStartOrSubmitResponse<CaseData, State> midEvent(
         final CaseData data = details.getData();
         final List<String> errors = new ArrayList<>();
 
-        if (null != data.getCicCase() && null != data.getCicCase().getAddress()) {
+        if (null != data.getCicCase()
+            && data.getCicCase().getContactPreferenceType().equals(ContactPreferenceType.POST)
+            && null != data.getCicCase().getAddress()) {
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/sptribs/common/event/page/SubjectDetails.java:41]: Same as comment above.

**Review by osheaf01 — CHANGES_REQUESTED** [review]
Again LGTM, bar just one suggestion as to how to check enum equality.

File: src/main/java/uk/gov/hmcts/sptribs/ciccase/model/ContactPreferenceType.java [hunk 5]
```diff
@@ -9,14 +9,14 @@
...
     @JsonProperty("Email")
     EMAIL("Email", "Email"),
     @JsonProperty("Post")
     POST("Post", "Post");
+
     private String type;
     private final String label;
+
     public boolean isEmail() {
         return EMAIL.name().equalsIgnoreCase(this.name());
```
Comment (osheaf01) [inline src/main/java/uk/gov/hmcts/sptribs/ciccase/model/ContactPreferenceType.java:21]: This has the same problems as the isPost() method originally proposed - could we fix this as part of this change?
Or else raise a tech debt ticket...

**Review by muradalamv1 — APPROVED** [review]

**Review by osheaf01 — APPROVED** [review]
LGTM

**Pull request merged**